### PR TITLE
Add ListenableAsyncCloseable#onClosing()

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
It is useful to get a notified on the leading edge when closing beings. This may drive decisions like stopping to use a connection or resources that is being closed and not suitable for reuse.

Modifications:
- Use existing transport mechansims to drive onClosing for connections.
- Remove NettyConnectionContext override of onClosing and defer to the base class.

### CODE REVIEW

File: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/CloseUtils.java [hunk 1]
```diff
@@ -44,6 +44,9 @@ private CloseUtils() {
      * @param closingStarted a {@link CountDownLatch} to notify
      */
     static void onGracefulClosureStarted(ConnectionContext cc, CountDownLatch closingStarted) {
+        // cc.onClosing() will trigger on the leading edge of closure, which maybe when the user calls closeAsync().
```
Comment (Scottmitch) [inline servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/CloseUtils.java:47]: this was an interesting observation... if we needed to differentiate between "any trigger for closing" (which is what `onClosing` is) and "transport/protocol is closing" we could add another method in the future, but currently this seems to be mostly a concern in tests.

**Review by idelpivnitskiy — CHANGES_REQUESTED** [review]
Thanks for catching other cases when we miss overriding a `default` method!
Everything looks good. However, there are a couple more cases that need an override for `onClosing()`:
- AbstractNettyIoExecutor
- ClientFactory in generated gRPC code

Also, please apply this patch:
```patch
Index: servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/NettyChannelListenableAsyncCloseable.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/NettyChannelListenableAsyncCloseable.java b/servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/NettyChannelListenableAsyncCloseable.java
--- a/servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/NettyChannelListenableAsyncCloseable.java	(revision cfd0bf0cbf7d299d87076e714c32ec2c55ad1da9)
+++ b/servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/NettyChannelListenableAsyncCloseable.java	(date 1669842880284)
@@ -101,6 +101,7 @@
      *
      * @return a {@link Completable} that notifies when the connection has begun its closing sequence.
      */
+    @Override
     public final Completable onClosing() {
         return fromSource(onClosing);
     }
@@ -162,7 +163,7 @@
         return onClose;
     }
 
-    final Completable onCloseNoOffload() {
+    private Completable onCloseNoOffload() {
         return onCloseNoOffload;
     }
 ```

**Comment by Scottmitch** [comment]
> AbstractNettyIoExecutor

This was intentional as there isn't a reliable leading edge trigger for `EventLoopGroup`. I can add some state in this class to provide best effort leading edge notification though.

**Comment by idelpivnitskiy** [comment]
Thanks!

**Pull request merged**